### DISCUSSION TITLE
Need for a 'query user' spécial operation

### DISCUSSION THREAD

**Discussion by cminssie** [meerk40t/meerk40t#2462] [Ideas]
Hi, I use a basic K40 with M2 nano ctlr that cannot control laser power from software.
I often configure multiple operation that need different laser settings(for example 7mA for engraving and 10mA for cutting).  I would like to be able to chain the operations in one single phase with a dialogue box asking the operator to change the laser power beetween individual tasks : 'please change the laser power to 10mA and press any key to continue'. It will remember the operator the right mA setting and avoid him miss the change. 
It may be easy to implement as a spécial operation ?
Thank for your great job!

**Comment by jpirnay** [comment]
Right click on the first operation and 'Insert special operation' - "Insert interrupt":
![grafik](https://github.com/meerk40t/meerk40t/assets/2670784/73642c94-59d4-4d1e-88c3-6da02ae671a6)

This will cause an 'interrupt' between the operations where the user has to confirm to proceed, 
![grafik](https://github.com/meerk40t/meerk40t/assets/2670784/137c0bae-5a88-4510-a8c1-92b123405226)

And yes you can change the text, just double click on the operation
![grafik](https://github.com/meerk40t/meerk40t/assets/2670784/9f83ea03-af7d-4de7-a2a8-7d55449f49a0)

**Comment by jpirnay** [comment]
And btw you can control the laser power... Just set the power value in the engrave operations, it's a sofware controlled power control - it has limitations but works okay most of the time - just try it

**Comment by Sophist-UK** [comment]
When I am using Meerk40t, I press the laser test button and adjust the digital power display so that the current is 12mA. This is 100% power. I then set the power independently for each Meerk40t operation node independently (and for rasters I can also set the DPI for the operation, and for images I can set the DPI per image) and I can run them without adjusting the hardware power at all.

Not only does Meerk40t adjust the power automatically between operations without any intervention from me, but when I merge operations it cleverly changes power as needed for each SVG element it burns.

From my own experience, the Meerk40t Power functionality works extremely well - I set the hardware power for 12mA, and in almost all the burns I do I make the power adjustments using Meerk40t. The only exception is when I am burning materials that are so easily burned (e.g. tissue paper, thin craft foam) that the MK power setting would need to be so low that the laser would be off for a distance which is visible to the naked eye - and in these cases I need to use a lower hardware power.

Now some people state (without really understanding the technology in detail) that this isn't "true" or "genuine" power control, but in order to set everyone's mind at ease about this, here is how I believe the technology works...

GRBL etc. controllers adjust the power using Pulse Width Modulation - which means that they turn the power on and off at an extremely fast frequency (fixed frequency?), and they vary the width of the on and off values so that the average voltage/current varies.

AFAIK, the digital power display for the PSU on a K40 also uses PWM to adjust the power.

Meerk40t's power setting also turns the laser on and off very frequently, however it works slightly differently based on distance rather than time. It overlays this on top of the existing PSU PWM, but in essence the principle it is using to adjust the power is almost exactly the same as GRBL/Lightburn uses.

**Comment by cminssie** [comment]
All right, thank you both, this 'interrupt' spécial operation is what I was looking for. I must admit that I never tried to set the main  power of the laser using  the power property in  each operation instead of turning the knob beetween operations.
I will try for sure! 👍

**Comment by Sophist-UK** [comment]
@cminssie Well, you are in for a treat. Meerk40t is capable of pretty much anything you can think of.

Have you tried the optimisations, because they will reduce your burn times substantially? At a minimum you need to optimise within each operation, but merging operations is even better.

I probably should also tell you about my "power" philosophy - which (having set 12mA) is to try to burn using full power in MK, If I want to reduce the burn then I increase speed (or for rasters / images reduce the DPI). By increasing the speed rather than reducing the power, I finish burns much faster.

**Reply by cminssie** [reply]
On my side, I avoid too much speed as it leads to lot of vibrations on my K40. It can results in poor details and even wavy borders in engrave or cut mode. Maybe I should track why so much vibrations, but I don't have references to compare with others.. thanks for your advices.

**Reply by Sophist-UK** [reply]
1. Check that the feet are fully stable. (I replaced the stock lockable-wheels with fixed feet.)
2. Try changing the tension of your belts.
3. For vector operations (engrave or cut) the max speed depends on the amount of curvature of the paths (i.e. the acceleration needed for the head).

**Comment by tatarize** [comment]
Speed actually varies what is basically power too, amount of photons per unit area. Though really the software controlled power entirely works for the M2. You can certainly set interrupts to adjust the power but it's very likely to be overkill.
